core/src/main/java/org/infinispan/reactive/publisher/impl/PublisherHandler.java

```diff
@@ -0,0 +1,567 @@
+       * the next batch to be ready for the next request as it comes, which is submitted on the {@code cpuExecutor}.
+       * @return future that will contain the publisher response with the data
+       */
+      CompletableFuture<PublisherResponse> results() {
+         boolean submitRequest = false;
+         CompletableFuture<PublisherResponse> currentFuture;
+         synchronized (this) {
+            if (futureResponse == null) {
+               currentFuture = new CompletableFuture<>();
+               currentFuture.thenRunAsync(this, cpuExecutor);
+               futureResponse = currentFuture;
+            } else {
+               currentFuture = futureResponse;
+               futureResponse = null;
+               submitRequest = true;
+            }
+         }
+         if (submitRequest) {
+            // Handles closing publisher or requests next batch if not complete
+            // Note this is not done in synchronized block in case if executor is within thread
+            cpuExecutor.execute(this);
+         }
+         if (trace) {
+            log.tracef("Retrieved future %d for request id %s", System.identityHashCode(currentFuture), requestId);
+         }
+         return currentFuture;
+      }
+
+      /**
+       * This will either request the next batch of values or completes the request. Note the completion has to be done
+       * after the last result is returned, thus it cannot be eagerly closed in most cases.
+       */
+      @Override
+      public void run() {
+         if (trace) {
+            log.tracef("Running handler for request id %s", requestId);
+         }
+         if (!complete) {
+            int requestAmount = batchSize;
+            if (trace) {
+               log.tracef("Requesting %d additional entries for %s", requestAmount, requestId);
+            }
+            upstream.request(requestAmount);
+         } else {
+            synchronized (this) {
+               if (futureResponse == null) {
+                  closePublisher(requestId, this);
+               } else if (trace) {
+                  log.tracef("Skipping run as handler is complete, but still has some results for id %s", requestId);
+               }
+            }
+         }
+      }
+   }
+
+   /**
+    * Special PublisherState that listens also to what key generates a given set of values. This state is only used
+    * when keys must be tracked (EXACTLY_ONCE guarantee with map or flatMap)
+    * <p>
+    * The general idea is the publisher will notify when a key or entry is sent down the pipeline and we can view
+    * all values that result from that (assumes the transformations are synchronous). Thus we only send a result when
+    * we have enough values (>= batchSize) but also get to a new key. This means that we can actually return more
+    * values than the batchSize when flatMap returns more than 1 value for a given key.
+    */
+   class KeyPublisherState extends PublisherState implements Consumer<Object> {
```

Comment:
Don't pay too much attention to this inner class yet, it only works for map operations and doesn't work completely with flatMap. I should be able to fix it tomorrow however.